File: Gemini/Fáceis/radixsort/radixsort.py
# Input: How to implement the radixsort sort algorithm? in python

def radix_sort(arr):
    """Sorts an array of integers using Radix Sort.

    Args:
        arr: The array of integers to be sorted.

    Returns:
        The sorted array.
    """

    max_element = max(arr)
    exp = 1

    while max_element // exp > 0:
        counting_sort(arr, exp)
        exp *= 10

    return arr

def counting_sort(arr, exp):
    """Performs counting sort on an array based on a specific digit.

    Args:
        arr: The array to be sorted.
        exp: The exponent determining the digit to be considered.
    """

    n = len(arr)
    count = [0] * 10
    output = [0] * n

    for i in range(n):
        count[(arr[i] // exp) % 10] += 1

    for i in range(1, 10):
        count[i] += count[i - 1]

    for i in range(n - 1, -1, -1):
        index = (arr[i] // exp) % 10
        output[count[index] - 1] = arr[i]
        count[index] -= 1

    for i in range(n):
        arr[i] = output[i]

File: Gemini/Fáceis/radixsort/Radixsort.java
// Input: How to implement the radixsort sort algorithm? in Java

import java.util.Arrays;

public class RadixSort {
    public static void radixSort(int[] arr) {
        int max = Arrays.stream(arr).max().getAsInt();
        int exp = 1;

        while (max / exp > 0) {
            countSort(arr, exp);
            exp *= 10;
        }
    }

    private static void countSort(int[] arr, int exp) {
        int n = arr.length;
        int[] output = new int[n];
        int[] count = new int[10];

        for (int i = 0; i < n; i++) {
            count[(arr[i] / exp) % 10]++;
        }

        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }

        for (int i = n - 1; i >= 0; i--) {
            output[count[(arr[i] / exp) % 10] - 1] = arr[i];
            count[(arr[i] / exp) % 10]--;
        }

        System.arraycopy(output, 0, arr, 0, n);
    }

    public static void main(String[] args) {
        int[] arr = {170, 45, 75, 90, 802, 24, 2, 66};
        radixSort(arr);
        System.out.println(Arrays.toString(arr));
    }
}
